awx/main/scheduler/task_manager_models.py

```diff
@@ -139,3 +254,56 @@ def get_instance_groups_from_task_cache(self, task):
             logger.warn(f"No instance groups in cache exist, defaulting to global instance groups for task {task}")
             return task.global_instance_groups
         return igs
+
+
+class TaskManagerModels:
+    def __init__(self, **kwargs):
+        # We want to avoid calls to settings over and over in loops, so cache this information here
+        kwargs['control_task_impact'] = kwargs.get('control_task_impact', settings.AWX_CONTROL_NODE_TASK_IMPACT)
+        kwargs['controlplane_ig_name'] = kwargs.get('controlplane_ig_name', settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME)
+        self.instances = TaskManagerInstances(**kwargs)
+        self.instance_groups = TaskManagerInstanceGroups(task_manager_instances=self.instances, **kwargs)
+
+    @classmethod
+    def init_with_consumed_capacity(cls, **kwargs):
+        tmm = cls(**kwargs)
+        tasks = kwargs.get('tasks', None)
+
+        if tasks is None:
+            instance_group_queryset = kwargs.get('instance_groups_queryset', None)
+            # No tasks were provided, so we will fetch them from the database
+            task_status_filter_list = kwargs.get('task_status_filter_list', ['running', 'waiting'])
+            task_fields = kwargs.get('task_fields', ('task_impact', 'controller_node', 'execution_node', 'instance_group'))
+            from awx.main.models import UnifiedJob
+
+            if instance_group_queryset is not None:
+                logger.debug("******************INSTANCE GROUP QUERYSET PASSED -- FILTERING TASKS ****************************")
+                # Sometimes things like the serializer pass a queryset that looks at not all instance groups. in this case,
+                # we also need to filter the tasks we look at
+                tasks = UnifiedJob.objects.filter(status__in=task_status_filter_list, instance_group__in=[ig.id for ig in instance_group_queryset]).only(
+                    *task_fields
+                )
```

Comment:
@AlanCoding FYI @akus062381 found a bug in this when we were making requests to /api/v2/job_templates/<ID>/instance_groups, the serializer calls this with a instance group queryset already formed. So you could get a 500 if there were jobs running on a DIFFERENT instance group that wasn't in the queryset, because later in `consume_capacity` we assumed all the instance groups on the tasks should be in the instance group mapping

Reply:
Oh, I get what you're saying about this but. Good find.

I'm on the fence about whether `instance_groups_queryset` is worth the effort we're putting in to maintain it. Think about the multiple types of optimizations. We have a single IG in a list view of IGs. The main optimization the serializer is trying to accomplish is to avoid re-computing capacity for every object in the list. Even if we incur 1 additional query, we can save those extra 20.

On the other hand, I see how we avoid tracking capacity for a _large number_ of instance groups if the user created a lot...

Reply:
yeah so alternative to this fix is to drop the instance_group_queryset argumetn and always compute everything for all groups and all tasks......

but there is a tiny part of me that is still dreaming of partitioning the task manager work on instance groups and making it where instances can only belong to 1 instance group :smiling_imp: